src/server/system_services/bucket_server.js

```diff
@@ -1956,23 +1959,45 @@ async function delete_bucket_replication(req) {
     await replication_store.instance().delete_replication_by_id(replication_id);
 }
 
-function validate_replication(req) {
+function validate_replication(req, db_repl_rules) {
     const replication_rules = req.rpc_params.replication_policy.rules;
     // num of rules in configuration must be in defined limits
     if (replication_rules.length > config.BUCKET_REPLICATION_MAX_RULES ||
         replication_rules.length < 1) throw new RpcError('INVALID_REPLICATION_POLICY', 'Number of rules is invalid');
 
     const rule_ids = [];
     const pref_by_dst_bucket = {};
+    const src_bucket = req.system.buckets_by_name && req.system.buckets_by_name[req.rpc_params.name.unwrap()];
 
     for (const rule of replication_rules) {
         const { destination_bucket, filter, rule_id } = rule;
+
         const dst_bucket = req.system.buckets_by_name && req.system.buckets_by_name[destination_bucket.unwrap()];
         // rule's destination bucket must exist and not equal to the replicated bucket
         if (req.rpc_params.name.unwrap() === destination_bucket.unwrap() || !dst_bucket) {
             throw new RpcError('INVALID_REPLICATION_POLICY', `Rule ${rule_id} destination bucket is invalid`);
         }
 
+        // validation for bidirectional replication - blocking bidirectional replication only for matching prefixes
+        if (destination_bucket.unwrap() !== req.rpc_params.name.unwrap()) {
+            const prefix = (filter && filter.prefix) || '';
+            // check if there already a rule consisting of src_bucket as destination_bucket for matching prefix
+            for (const db_rules of db_repl_rules) {
+                if (_.isEqual(db_rules._id, dst_bucket.replication_policy_id)) {
+                    const matching_rule = db_rules.rules.find(
+                        db_rule =>
+                            _.isEqual(src_bucket._id, db_rule.destination_bucket) &&
+                            (!db_rule.filter || db_rule.filter.prefix.toString().startsWith(prefix) ||
+                             prefix.toString().startsWith(db_rule.filter.prefix.toString()))
```

Comment:
we should also block if one way contain empty prefix.

Reply:
Not sure, if I understand your question but this condition will check for empty prefix as well. Also, I have written a test case for the same.

Reply:
True, and great adding the tests.